src/Models/HydrostaticFreeSurfaceModels/SplitExplicitFreeSurfaces/split_explicit_free_surface.jl

```diff
@@ -0,0 +1,318 @@
+end
+
+@inline   cosine_averaging_kernel(τ::FT) where FT = τ ≥ 0.5 && τ ≤ 1.5 ? convert(FT, 1 + cos(2π * (τ - 1))) : zero(FT)
+@inline constant_averaging_kernel(τ::FT) where FT = convert(FT, 1)
+
+""" An internal type for the `SplitExplicitFreeSurface` that allows substepping with
+a fixed `Δt_barotropic` based on a CFL condition """
+struct FixedTimeStepSize{B, F}
+    Δt_barotropic    :: B
+    averaging_kernel :: F
+end
+
+""" An internal type for the `SplitExplicitFreeSurface` that allows substepping with
+a fixed number of substeps with time step size of `fractional_step_size * Δt_baroclinic` """
+struct FixedSubstepNumber{B, F}
+    fractional_step_size :: B
+    averaging_weights    :: F
+end
+
+function FixedTimeStepSize(grid;
+                           cfl = 0.7,
+                           averaging_kernel = averaging_shape_function,
+                           gravitational_acceleration = g_Earth)
+
+    FT = eltype(grid)
+
+    Δx⁻² = topology(grid)[1] == Flat ? 0 : 1 / minimum_xspacing(grid)^2
+    Δy⁻² = topology(grid)[2] == Flat ? 0 : 1 / minimum_yspacing(grid)^2
+    Δs   = sqrt(1 / (Δx⁻² + Δy⁻²))
+
+    wave_speed = sqrt(gravitational_acceleration * grid.Lz)
+
+    Δt_barotropic = convert(FT, cfl * Δs / wave_speed)
+
+    return FixedTimeStepSize(Δt_barotropic, averaging_kernel)
+end
+
+@inline function weights_from_substeps(FT, substeps, averaging_kernel)
+
+    τᶠ = range(FT(0), FT(2), length = substeps+1)
+    Δτ = τᶠ[2] - τᶠ[1]
+
+    averaging_weights = map(averaging_kernel, τᶠ[2:end])
+    idx = searchsortedlast(averaging_weights, 0, rev=true)
+    substeps = idx
+
+    averaging_weights = averaging_weights[1:idx]
+    averaging_weights ./= sum(averaging_weights)
+
+    return Δτ, tuple(averaging_weights...)
+end
+
+Base.summary(s::FixedTimeStepSize)  = string("Barotropic time step equal to $(prettytime(s.Δt_barotropic))")
+Base.summary(s::FixedSubstepNumber) = string("Barotropic fractional step equal to $(s.fractional_step_size) times the baroclinic step")
+
+Base.summary(sefs::SplitExplicitFreeSurface) = string("SplitExplicitFreeSurface with $(summary(sefs.substepping))")
+
+Base.show(io::IO, sefs::SplitExplicitFreeSurface) = print(io, "$(summary(sefs))\n")
+
+#####
+##### Maybe extend halos in Connected topologies
+#####
+
+const ConnectedTopologies = Union{LeftConnected, RightConnected, FullyConnected}
+
+# Extending halos is not allowed with variable time-stepping
+function maybe_extend_halos(TX, TY, grid, ::FixedTimeStepSize) 
+
+    if TX() isa ConnectedTopologies || TY() isa ConnectedTopologies
+        throw(ArgumentError("A variable substepping through a CFL condition is not supported for the `SplitExplicitFreeSurface` on $(summary(grid)). \n
+                             Provide a fixed number of substeps through the `substeps` keyword argument as: \n
+                             `free_surface = SplitExplicitFreeSurface(grid; substeps = N)` where `N::Int`"))
```

Comment:
This is another example of an error that should be in the constructor. Sprinkling the errors / unsupported configurations throughout the code makes it hard to follow the logic of the constructors and user interface.